### ISSUE TITLE
Improve prohibits command to list installed packages that prevent a package from being installed due to that package's requires

### ISSUE THREAD

**Issue by peterrehm** [composer/composer#5190]
With the following composer.json:

```json
{
    "require-dev" : {
        //...
        "sebastian/phpcpd": "^2.0"
    }
}
```

When I run this command:

```
composer update
```

I get this output:

Only sebastian/phpcpd (2.0.1) is installed.

And I expected this to happen:

sebastian/phpcpd (2.0.2) should be installed.

@enygma was reporting the same issue through twitter.

**Comment by curry684** [comment]
Just tried to reproduce:
```sh
$ composer require sebastian/phpcpd:^2.0
./composer.json has been created
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing theseer/fdomdocument (1.6.1)
    Downloading: 100%         

  - Installing phpunit/php-timer (1.0.7)
    Loading from cache

  - Installing symfony/polyfill-mbstring (v1.1.1)
    Loading from cache

  - Installing symfony/console (v2.8.4)
    Downloading: 100%         

  - Installing sebastian/version (1.0.6)
    Loading from cache

  - Installing symfony/finder (v3.0.4)
    Loading from cache

  - Installing sebastian/finder-facade (1.2.1)
    Downloading: 100%         

  - Installing sebastian/phpcpd (2.0.2)
    Downloading: 100%         

symfony/console suggests installing symfony/event-dispatcher ()
symfony/console suggests installing symfony/process ()
symfony/console suggests installing psr/log (For using the console logger)
Writing lock file
Generating autoload files
$ cat composer.json 
{
    "require": {
        "sebastian/phpcpd": "^2.0"
    }
}
```
Seems to work fine, getting `2.0.2`. Do you have detailed steps to reproduce from an empty repository?

**Comment by peterrehm** [comment]
Hrm, you are right it can not be reproduced in isolation. That is strange. I will try to figure out if it is dependency related.

**Comment by peterrehm** [comment]
After further investigation I am still having that issue in the project I am working on, I am closing the issue for now, reopening once I have more details. I still can't get version 2.0.2 in that project on others the same constraint is fine.

````json
$ composer prohibits sebastian/phpcpd
xxxx/yyyyy  branch  requires (for development)  sebastian/phpcpd (^2.0)
````

**Comment by curry684** [comment]
Try `$ composer prohibits sebastian/phpcpd 2.0.2` instead  :)

**Comment by peterrehm** [comment]
No further insights:

````
$ composer prohibits sebastian/phpcpd 2.0.2
There is no installed package depending on "sebastian/phpcpd" in versions not matching 2.0.2
````

**Comment by curry684** [comment]
Hmm that's really weird - that indicates an upgrade should most certainly pick it up. If you have a `composer.json` and possibly partner `composer.lock` that allows me to reproduce it I'll happily look into it.

**Comment by Seldaek** [comment]
Just look at the requirements of sebastian/phpcpd between 2.0.1 and 2.0.2 they changed from `>=...` to `~...`, so suddenly higher versions are not allowed anymore:

```
sebastian/finder-facade: ~1.1
sebastian/version: ~1.0
symfony/console: ~2.2
phpunit/php-timer: ~1.0
```
Most likely you're using symfony 3 and that's blocking it, or something like that.

**Comment by curry684** [comment]
Hmm well spotted, and I can reproduce it with `symfony/symfony:^3.0` in place, but that means there's a bug in `prohibits` there.

**Comment by Seldaek** [comment]
Well nothing prevents it from being installed, but its own requirements prevent it from being installed.

**Comment by Seldaek** [comment]
So yeah if a specific version is given like that, maybe we should check if any of its requirements are not satisfied by current installed packages and if so list these as well.

**Comment by curry684** [comment]
I just looked at a quick fix for this, but the Symfony use case makes it even messier than it already should be - as `phpcpd:2.0.2` depends on `symfony/console:~2.2`, but my own project requires `symfony/symfony:^3.0` causing `symfony/console:3.0.4` to appear via its replacements.

Sensibly detecting cases like that without resorting to a full resolve is going to be hard to implement at the very least. Not impossible, but complex.

**Comment by stof** [comment]
@curry684 it is not messier. You just have to check for replacements, which is something that composer is already able to do

**Comment by peterrehm** [comment]
@Seldaek @curry684 Thanks for the awesome support!

**Comment by curry684** [comment]
@stof it changes a triple-nested foreach into a 5 or 6 nested foreach to check all options, that's kinda messy hehe. Just a word.

**Comment by Seldaek** [comment]
OK I used this to force the case at hand (2.0.3 was since released fixing the requirements of phpcpd):

```
{
    "require": {
        "symfony/symfony": "3.*"
    },
    "require-dev" : {
        "sebastian/phpcpd": "^2.0,<2.0.3"
    }
}
```

And with the fix in b6680b6f2ab831f3918ca3be42d5cb2594f830ad I get this:

```
$ c why-not sebastian/phpcpd:2.0.2
sebastian/phpcpd  2.0.2       requires          sebastian/version (~1.0)
__root__          dev-master  does not require  sebastian/version (but 2.0.0 is installed)
sebastian/phpcpd  2.0.2       requires          symfony/console (~2.2)
__root__          dev-master  requires          symfony/symfony (3.*)
```

So it shows sebastian/version has a difference but that's a false positive in this case, still showing it because it might be relevant. And then it shows that the requires for symfony are clearly different so that would let us solve it quickly I believe.

Other versions:

```
$ c why-not sebastian/phpcpd:2.0.3
__root__          dev-master  requires (for development)  sebastian/phpcpd (^2.0,<2.0.3)
sebastian/phpcpd  2.0.3       requires                    sebastian/version (~1.0)
__root__          dev-master  does not require            sebastian/version (but 2.0.0 is installed)

$ c why-not sebastian/phpcpd:2.0.4
__root__  dev-master  requires (for development)  sebastian/phpcpd (^2.0,<2.0.3)
```

As you see 2.0.3 shows us why it's not installed due to the root, but also because it has other requirements that don't match what's installed. While the sebastian/version requirement was fixed again in 2.0.4 and it disappears then.

**Comment by peterrehm** [comment]
Awesome, thanks!
